Control node enhancements

groud opened the pull request:
API changes:
- [x] Replace anchor mode by a float between 0 and 1. By now, it stay modifiable through an enum property, and the ANCHOR_BEGIN, ANCHOR_END and ANCHOR_CENTER are kept as constant. Basically it does not deeply changes the API or the workflow, but it stores the anchor as a float value. (This is the base for other enhancements)
- [x] The margin are now always a vector2 relative to the anchor position, it does not changes sign when it has a value like ANCHOR_CENTER or ANCHOR_END (>=0.5). This means that inner margins on the bottom and right side (with ANCHOR_CENTER or ANCHOR_END anchor) have usually negative values. It explains the first commit size.

Editor:
- [x] Anchor visualization,
- [x] Draggable anchors,
- [x] Automatically snapping anchors depending on the zoom level.
- [x] Draw anchors lines when dragged, which are then colored when snapped.
- [x] display percentage when draggin anchors.
- [x] snap anchors one to each other.
- [x] drag several anchors at the same time.
- [x] Avoid left anchor to over right

Abandoned:
~~highlight the side related to the dragged anchor~~ (I finally think this is useless, but the link beween anchors and the control sides could be made when moving control nodes, with a display of the margins values when dragging it)

Demo:
![output](https://user-images.githubusercontent.com/6093119/28989639-fe1afc2a-7976-11e7-80cb-fa8d0d70bcc1.gif)

(the green pins are anchors, sorry I could not capture the mouse)

Any other ideas ? What do you think ?

groud committed:
Replace GUI anchor type by a float between 0 and 1

groud committed:
Add anchor visualization

reduz commented:
Was this not proposed before?

On Wed, Jul 26, 2017 at 8:09 PM, Gilles Roudiere <notifications@github.com>
wrote:

> API changes:
>
>    - Replace anchor mode by a float between 0 and 1. By now, it stay
>    modifiable through an enum property, and the ANCHOR_BEGIN, ANCHOR_END and
>    ANCHOR_CENTER are kept as constant. Basically it does not deeply changes
>    the API or the workflow, but it stores the anchor as a float value. (This
>    is the base for other enhancements)
>    - The margin are now always a vector2 relative to the anchor position,
>    it does not changes sign when it has a value like ANCHOR_CENTER or
>    ANCHOR_END (>=0.5). This means that inner margins on the bottom and right
>    side (with ANCHOR_CENTER or ANCHOR_END anchor) have usually negative
>    values. It explains the first commit size.
>
> Editor:
>
>    - Anchor visualization (still buggy with rotated nodes). Preview :
>    [image: 2017-07-27-005244_2200x1098_scrot]
>    <https://user-images.githubusercontent.com/6093119/28647484-c8c7ffde-7267-11e7-9d1e-c15630e5d7ff.png>
>    (the green pins are anchors)
>
> To be done soon:
>
>    - draggable anchors
>    - automatically-snapping anchor
>    - display percentage when draggin anchors ?
>    - highlight the side related to the dragged anchor ?
>
> Any other ideas ? What do you think ?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/godotengine/godot/pull/9889
> Commit Summary
>
>    - Replace GUI anchor type by a float between 0 and 1
>    - Add anchor visualization
>
> File Changes
>
>    - *M* doc/base/classes.xml
>    <https://github.com/godotengine/godot/pull/9889/files#diff-0> (26)
>    - *M* editor/call_dialog.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-1> (20)
>    - *M* editor/code_editor.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-2> (8)
>    - *M* editor/editor_name_dialog.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-3> (2)
>    - *A* editor/icons/2x/icon_editor_control_anchor.png
>    <https://github.com/godotengine/godot/pull/9889/files#diff-4> (0)
>    - *A* editor/icons/icon_editor_control_anchor.png
>    <https://github.com/godotengine/godot/pull/9889/files#diff-5> (0)
>    - *A* editor/icons/source/icon_editor_control_anchor.svg
>    <https://github.com/godotengine/godot/pull/9889/files#diff-6> (98)
>    - *M* editor/output_strings.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-7> (6)
>    - *M* editor/plugins/animation_tree_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-8> (12)
>    - *M* editor/plugins/camera_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-9> (2)
>    - *M* editor/plugins/canvas_item_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-10> (41)
>    - *M* editor/plugins/canvas_item_editor_plugin.h
>    <https://github.com/godotengine/godot/pull/9889/files#diff-11> (4)
>    - *M* editor/plugins/multimesh_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-12> (2)
>    - *M* editor/plugins/sample_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-13> (6)
>    - *M* editor/plugins/sample_library_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-14> (4)
>    - *M* editor/plugins/sample_player_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-15> (14)
>    - *M* editor/plugins/spatial_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-16> (4)
>    - *M* editor/project_settings_editor.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-17> (4)
>    - *M* editor/property_editor.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-18> (14)
>    - *M* main/tests/test_gui.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-19> (4)
>    - *M* modules/gridmap/grid_map_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-20> (2)
>    - *M* modules/visual_script/visual_script_editor.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-21> (2)
>    - *M* scene/gui/control.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-22> (111)
>    - *M* scene/gui/control.h
>    <https://github.com/godotengine/godot/pull/9889/files#diff-23> (27)
>    - *M* scene/gui/dialogs.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-24> (4)
>    - *M* scene/gui/graph_edit.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-25> (4)
>    - *M* scene/gui/item_list.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-26> (6)
>    - *M* scene/gui/popup.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-27> (18)
>    - *M* scene/gui/rich_text_label.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-28> (2)
>    - *M* scene/gui/scroll_container.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-29> (4)
>    - *M* scene/gui/tab_container.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-30> (12)
>    - *M* scene/main/viewport.cpp
>    <https://github.com/godotengine/godot/pull/9889/files#diff-31> (4)
>
> Patch Links:
>
>    - https://github.com/godotengine/godot/pull/9889.patch
>    - https://github.com/godotengine/godot/pull/9889.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/9889>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2w39J9XfWav3iG1PR5FjzCQ7ympMks5sR8dEgaJpZM4OkmYa>
> .
>

groud commented:
Yes sorry, I should have mention that. Here: #9527 
The point is that it is made here so that it does not change (by now) the way we work with anchors.
The previous proposal was refused for not being user friendly. This is no more the case.

As I want to work more on that, I created the pull request soon enough so that my later changes (anchor draggint, etc...) do not break compatibility.

reduz commented:
I would probably make it user friendly like this, the enum could be:

ANCHOR_BEGIN, ANCHOR_END, ANCHOR_CENTER, ANCHOR_RATIO

you still let them exist as is, and not change anything internally, save
for adding a function:

set_anchor_ratio(margin,ratio) which is ignored unless mode is set to
RATIO, but does not make the existing API or UI different.

then when ANCHOR_RATIO is enabled, you make visible a new property below
the anchor, eg "Ratio Left" that goes from 0 to 1.

This way, this functionality, which is clearly a corner case, is only
present when the right mode is selected, otherwise you can easily ignore
the fact it exists and does not get in the way.

That said, I am still not sure if the functionality itself is desired. Do
you have any examples of what could be done with this that, in the worst
case, can't be done with containers?


On Thu, Jul 27, 2017 at 3:39 AM, Gilles Roudiere <notifications@github.com>
wrote:

> Yes sorry, I should have mention that. Here: #9527
> <https://github.com/godotengine/godot/pull/9527>
> The point is that it is made here so that it does not change (by now) the
> way we work with anchors.
> The previous proposal was refused for not being user friendly. This is no
> more the case.
>
> As I want to work more on that, I created the pull request soon enough so
> that my later changes (anchor draggint, etc...) do not break compatibility.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/9889#issuecomment-318273045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2zPsbNXT79v5R48TqRvSWSJhhwvaks5sSDCxgaJpZM4OkmYa>
> .
>

groud commented:
Well, floating anchors is a common GUI design. The problem is that in its current implementation it is not perfectly flexible, and needs containers to achieve all possible GUI. (With floating anchors we could, for example, recreate the behaviour of a BoxContainer.)

The fact is that godot mixes two kind of paradigms in the GUI design. Adding this would allow people that prefer to the anchor/margin paradigm to only use that, while other that prefer the container paradigm could still use it.
To spot the differences between the two paradigms:
- In the anchor/margin paradigm the size and position of a node are handled by the child node.
- In the container paradigm the  the size and position are handled by the parent.
I think both paradigm should be usable "alone" (and should be more integrated one with the other, but that's another problem).

In any case, I think all this ANCHOR_* is not clear at all when you start playing with the GUI nodes. In terms of discoverability its really hard to understand: the margins change sign when you change the anchor, the anchors are not displayed, etc. i had myself problems understanding what they were used for.

In my opinion, we should keep the anchor as a float value (remove the enum), and make people use either draggable anchors instead, or the anchor menu at the top of the 2D editor to move the anchors. The property editor would not be used a lot anymore, and would make clearer the whole thing.

Noshyaar commented:
>Do you have any examples of what could be done with this that, in the worst case, can't be done with containers?

Container alone cannot keep user-defined constant ratio between its children.

HBoxContainer with two children in EXPAND_FILL, each child *will* take exactly half of the parent's size, or whatever set as minimum size if it's larger, which is predefined.

What if I want one side bigger than another, like, exactly 1:2? (maybe right side has more data)

You can't, at least without coding.

With ANCHOR_RATIO (in older Godot), I can set left child to end at/right child to start at 0.33.

groud commented:
Thanks @Noshyaar , this is quite hard to explain, you did it well.
When I am able to drag the anchors, I think I will make a video explaining how it can be used, so that everyone could understand how it works. :)

groud committed:
Make anchors draggable

groud committed:
Snap anchors when dragged, depending on the zoom level

groud committed:
Add lines to make understandable when anchors are snapped

groud commented:
Anchors are now draggable, they snap to the begin/center/end and lines are here to show where they are and when they are snapped.
See the updated demo in the first post.

groud committed:
Simplifies the canvas editor code

groud committed:
Displays percentages when dragging anchors

groud committed:
Make anchors snap to each other

groud commented:
Updated with the display of several percentages when draggin anchors.
Also, anchors now snap to each other (left anchor snaps to right and top to bottom, and vice-versa)

groud committed:
Added the possibility to move all anchors at once when they are clustered

toger5 commented:
How do you display if the anchor is a end or begin anchor?
Could u use dotted lines to indicate which distances will change on resize of the parent?
The anchoring system in godot is really good. the only issue is that the input with `margin` and `begin end` is a little confusing. Your solution is much better!

groud commented:
Well the anchor is no more either `begin` or `end`, it is here a value between 0 and 1 (with 0 corresponding to begin and 1 to end). So basically, when you drag the anchors to the top (or left) of the parent it corresponds to the previous begin value and if you drag them to bottom (or right) to the end value.

You can have a look to Unity's rect transform reference to understand what I want to achieve:
https://www.youtube.com/watch?v=FeheZqu85WI

> Could u use dotted lines to indicate which distances will change on resize of the parent?

You mean instead of the plain gray lines ? The fact is that the current canvasitem API does not allow for such draw easily. It might require a special built-in shader, but that's out of the scope of this PR.

> The anchoring system in godot is really good. the only issue is that the input with margin and begin end is a little confusing. Your solution is much better!

Thank's, can't disagree with that ^^


Aside from that, I added the possibility to drag all four anchors at once when clicking on the center of the four anchor. This really makes easy moving them from one point to another.
A little more work and it should be ready to merge ! :D

toger5 commented:
I think I got it:
so basically the margin is the distance between the line of two anchor marks and the edge of the control:
topMargin is the distance between the line of topLeft and the topRight anchor and the top part of the control node and stays constant on resizing.
the anchor points don't have constant positions instead their positions are percentage based in the parent.

groud commented:
Well it's more like (for the X axis):
- the anchor is relative to the parent, 0% being at the extreme left of the parent control node and 100% at it's right (it thus depends on the parent size).
- the margin is a fixed pixel-based offset of the side of the child node from the anchor.

toger5 commented:
That is how I understood it. And the more I think about it there more I like it!
you are right it's almost like it was before just makes even more sense.

Also adding distance display for the anchor to control relation (the margin) would be great.

groud commented:
I guess the only subtle difference will be the fact that the margin will be allways positive when the node side is at the right/bottom of the corresponding floating anchor. In the previous version, when the anchor was placed at the end (=right of the parent node), a positive value for the margin positioned the node at the left the anchor. This was a little bit confusing, as margins kept changin sign when changing the anchor mode... 
Also, this explains why I had to change some margin values into the code, from positive to negative. :)
Indeed, it's common to have the child node inside the parent node, so when the anchors is at the right, the margin is usually negative.

toger5 commented:
Interesting,
yea I saw the changes you made to margins.
Are you sure this is the best way to go with? Because it will be hard to change ALL the right and bottom margins. That are also used a lot in drawing functions: `get_margin(MARGIN_RIGHT)`...
What about all the style boxes do they also need to change their usage of margins?
Also I usually just go to a node and than type in the same value for all the margins and expect it to be inside the parent node.

Also web kind of sets the standard how to think of margins. and there you also don't use signing.

If it is fairly easy to define margin left as positive x and margin right as negative x relative to their anchors might be better.

I think I see the issue though:
when both anchors are at the left, than I would need to use a negative value for margin right and a positive for margin left...
but honestly I think this is still better... because with the way you display the anchors it kind of makes sense when the I anchor to the left edge, that I need to use negative for the right margin because it is outside of the anchor rect?
Another idea would be to (as right now) swap the signing when over/under 50%.
Although this also would lead to a LOT of confusion...

groud committed:
Handle undo/redo of anchor positioning

groud commented:
Well the margin are only changed for the control node, not for the theming :). The margins for style boxes are the same.

In my first implementation I tried with a sign swap at 50%, but yeah... this is confusing. I think the current proposal need an habit change, but I think that in terms of discoverability/understandability this is a non neglictible improvement.
Indeed, the margin keep their sign, independantly of the anchor position. I mean, elseway it makes no sense to keep the anchor value at 0 on the left and 1 on the right for the right side: anchors values could be swapped for the right side (0 to right, 1 to left) if we also swap the margins. At least here we keep all positioning values increasing in the same direction: decreasing a value (anchor or margin) move the node to the left while increasing it moves it to the right: this is consistent and easily understandable.

To be honest, I think that in this situation the term "margin" should be replaced by something like "offset". This could reduce the confusion, but this can be discussed later on.

Also, from what I tested by now, the PR as is does not seem to create any problems, at least in the editor. but maybe in some specific situations signs problems could be still left.

toger5 commented:
Renaming to offset (anchor offset) is a very good approach. At the beginning i head issues understnding all the different marigins. Stylebox margin, default margin... Offset at least makes it clear, that they are sth different.
I didnt check your whole pr. But i thibk ther will be a lot of drawing functions using the stylebox margins. My proposal would be to add a get offset function to the stylebox class which than return negative values for right and bottom... Than it would just be a matter of replacing get_marfin with get_offset wherever it is used in conjunction with offset

groud commented:
Why not. I think this is a little bit out of the scope of this PR, but those functions could be helpful to draw the styleboxes with this new system. :)
For now I need more opinions on renaming `margin` to `offset`. :)

groud commented:
> Are you sure this is the best way to go with? Because it will be hard to change ALL the right and bottom margins. That are also used a lot in drawing functions: get_margin(MARGIN_RIGHT)...
To add to the conversation, I checked my code. I believe I changed all the code using the get_margin from a control node so that it takes into account the new way of handling anchors. So it should note cause any problem :)

groud commented:
In any case I think it's ready to merge now :). Ping @akien-mga and maybe @RandomShaper and @bojidar-bg (found you in the contributors of control.cpp and canvas_item_editor_plugin.cpp).
For those who are not convinced yet, do not hesitate to retreive and compile the code from this PR. I think t's a cool improvement regarding usability :)

reduz commented:
this will warrant nice discussion on PR monday :P

groud commented:
Does this PR needs any more work before being merged ?

bojidar-bg commented:
@groud I used swapped anchors with -1, -1 scale. No matter how you set the other options, you won't be able to set it take the full size of the container without explicitly writing that size as margins. (a bit hard to explain)

Here is the screenshot:
![https://pasteboard.co/GEyEQ9D.png](https://cdn.pbrd.co/images/GEyEQ9D.png)

groud commented:
Ok, I see your point. Clearly reverted anchors do not work.
I'll have to fix this :)

groud commented:
Ok. The problem is in fact deeper than this PR. 
Indeed, this anchor-based design does not work well when the child node is either scaled or rotated relatively to its parent. This explains why, in Unity, they display a rectangle behind the real node when the child node is rotated relatively to its parent. The rectangle behind being drawn without any rotation/scale applied to it. 
I'll take a 45° rotated node as an example: there's no way to properly resize it accordingly to its parents without affecting two of its sides: 
![output](https://user-images.githubusercontent.com/6093119/29094511-dc72702a-7c8d-11e7-87e2-715ac71e93cd.gif)
(This does not even need reverted anchors :/ )

Consequently, I think we should stuck to the common use case, and not consider corner ones. Those should be handled with scripting or something like that. Instead, I'll avoid anchors to be reverted, And maybe add the rectangle draw behind as in unity.

Note that the anchors still work if only the parent is rotated/scaled:
![output](https://user-images.githubusercontent.com/6093119/29094718-984c2fac-7c8e-11e7-80ea-f993597f3e64.gif)

reduz commented:
anchors are not really supposed to work with transformed nodes, so pay that
no atention. It doesn't really make sure to use those together. Also,
transformation is applied after anchoring, so they don't really play
against each other.

On Tue, Aug 8, 2017 at 6:09 PM, Gilles Roudiere <notifications@github.com>
wrote:

> Ok. The problem is in fact deeper than this PR.
> Indeed, this anchor-based design does not work well when the child node is
> either scaled or rotated relatively to its parent. This explains why, in
> Unity, they display a rectangle behind the real node when the child node is
> rotated relatively to its parent. The rectangle behind being drawn without
> any rotation/scale applied to it.
> I'll take a 45° rotated node as an example: there's no way to properly
> resize it accordingly to its parents without affecting two of its sides:
> [image: output]
> <https://user-images.githubusercontent.com/6093119/29094511-dc72702a-7c8d-11e7-87e2-715ac71e93cd.gif>
> (This does not even need reverted anchors :/ )
>
> Consequently, I think we should stuck to the common use case, and not
> consider corner ones. Those should be handled with scripting or something
> like that. Instead, I'll avoid anchors to be reverted, And maybe add the
> rectangle draw behind as in unity.
>
> Note that the anchors still work if only the parent is rotated/scaled:
> [image: output]
> <https://user-images.githubusercontent.com/6093119/29094718-984c2fac-7c8e-11e7-80ea-f993597f3e64.gif>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/9889#issuecomment-321082668>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z20SLYHHB98kAPub2y3p5R8X2bfinks5sWM6bgaJpZM4OkmYa>
> .
>

reduz commented:
i mean it does not make sense to use them together, so it's not a problem.

On Tue, Aug 8, 2017 at 6:34 PM, Juan Linietsky <reduzio@gmail.com> wrote:

> anchors are not really supposed to work with transformed nodes, so pay
> that no atention. It doesn't really make sure to use those together. Also,
> transformation is applied after anchoring, so they don't really play
> against each other.
>
> On Tue, Aug 8, 2017 at 6:09 PM, Gilles Roudiere <notifications@github.com>
> wrote:
>
>> Ok. The problem is in fact deeper than this PR.
>> Indeed, this anchor-based design does not work well when the child node
>> is either scaled or rotated relatively to its parent. This explains why, in
>> Unity, they display a rectangle behind the real node when the child node is
>> rotated relatively to its parent. The rectangle behind being drawn without
>> any rotation/scale applied to it.
>> I'll take a 45° rotated node as an example: there's no way to properly
>> resize it accordingly to its parents without affecting two of its sides:
>> [image: output]
>> <https://user-images.githubusercontent.com/6093119/29094511-dc72702a-7c8d-11e7-87e2-715ac71e93cd.gif>
>> (This does not even need reverted anchors :/ )
>>
>> Consequently, I think we should stuck to the common use case, and not
>> consider corner ones. Those should be handled with scripting or something
>> like that. Instead, I'll avoid anchors to be reverted, And maybe add the
>> rectangle draw behind as in unity.
>>
>> Note that the anchors still work if only the parent is rotated/scaled:
>> [image: output]
>> <https://user-images.githubusercontent.com/6093119/29094718-984c2fac-7c8e-11e7-80ea-f993597f3e64.gif>
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/9889#issuecomment-321082668>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AF-Z20SLYHHB98kAPub2y3p5R8X2bfinks5sWM6bgaJpZM4OkmYa>
>> .
>>
>
>

groud commented:
@reduz, Maybe you can help me on that.
I wanted to register the ANCHOR_BEGIN, ANCHOR_CENTER, ANCHOR_END as constants with the BIND_CONSTANT() macro. (I obviously wanted them to be available to GDscript)
The problem is that BIND_CONSTANT registers the value as a integer, and i want to store a float value. :/
Is there any way I can do that ?

groud committed:
Adds a function to set Anchors with a layout preset

groud committed:
Avoids inverted anchors
Add a push_opposite_anchor argument pushing the opposite anchor if
needed

groud commented:
Rebased and:
- Added a new function (set_anchors_presets) to set anchors using a layout preset.
- Avoid that left/top anchors goes over the right/bottom one.
- Add a new optional arguments to the anchor to push the opposite anchor if necessary (else limits the anchors valu the the opposite's one) 

I think its ready to merge now :)

groud commented:
Ping @reduz  and @akien-mga

vnen commented on scene/gui/control.h:
```diff
@@ -51,11 +51,9 @@ class Control : public CanvasItem {
-		ANCHOR_END,
-		ANCHOR_CENTER,
-	};
+	static const float ANCHOR_BEGIN = 0.0f;
```
Getting an error with MSVC because of this: `error C2864: 'Control::ANCHOR_BEGIN': a static data member with an in-class initializer must have non-volatile const integral type`

vnen replied:
Same applies to the other `ANCHOR_` constants.

reduz replied:
This is also not C++03 compatible, needs to be moved to the CPP

groud replied:
Thanks for fixing :) 
Too bad I did not have this error with TravisCI.

Think-Jumper commented:
Hi，if you want to make a GUI design for godot, I have a advice for you that you could just try to use the FairyGUI: the 3rd UI editor for unity3d and some other engine. It's pretty good to use and create ui productive. It maybe can give you some inspirations and different views for making a GUI design for godot.
